docs: reference docs for remote builds (#778)
canonical/rockcraft: docs/reference/remote-build.rst

Author: https://canonical-ubuntu-documentation-library--778.com.readthedocs.build/rockcraft/en/778/reference/remote-build/

.. _ref-remote-build:

*************

> Remote builds

Author: @medubelko in the end I opted to document this in Rockcraft now, so that the release has the docs, and then worry about upstreaming this to craft-application in the near future. Once we review and merge this I can point to this new doc on the 1.7.0 release notes

*************

> Remote build is a feature in Rockcraft that offloads the build process to
> `Launchpad`_'s `build farm`_ to build multiple rocks simultaneously, and for
> different architectures than your running machine.

Reviewer: It's best not to put two links next to each other.

```suggestion
Remote builds offload rock builds to the `build farm`_ hosted by `Launchpad`_.
With remote builds, you can assemble multiple rocks simultaneously and build
for all supported architectures.
```

> Pre-requisites
> --------------

Reviewer: ```suggestion
Prerequisites
-------------
```

In order to perform remote builds, the following conditions must be met:

- You must have a `Launchpad account`_, as the remote builds are performed on
  Launchpad.
- The Rockcraft project must be version-controlled by Git. This is because
  Rockcraft uses a Git-based workflow to upload the project to Launchpad.
- The repository hosting the Rockcraft project must not be a shallow clone,
  because Git does not support pushing shallow clones.

> Overview

Reviewer: How about "How it works" or simply "Remote build process"?

I don't think the content quite fits the definition of overview, and typically an overview would come before prerequisites.

Author: I see your point. I think a section like "How it works" and "Remote build process" feels a bit... generic and redundant with the page itself, so I removed this section and put this paragraph in the "introduction", before the Prerequisites section. Looks like it might flow better like this.

Let me know what you think!

Reviewer: Looks good. :)

--------

Remote builds are launched by running ``rockcraft remote-build``. Rockcraft will
upload the Git repository on the current working directory to Launchpad on your
behalf, under your account. Next, it will trigger builds for the Rockcraft
project present on the root of the repository and continuously monitor the
status of the new builds.

Once all builds are done (either through a successful build or a failure), the
rock files will be downloaded to the current directory, together with the build
logs.

Limitations
-----------

The following is a list of the current limitations of the remote build feature,
which are planned to be addressed in the future:

> - The prospective rock must be open source and public, because the remote builds
>   will be publicly available.

Reviewer: ```suggestion
- The prospective rock must be open source and public, because remote builds
  are publicly available.
```

Author: there's a subtle difference here though: it's not that "all" remote builds are public, just the ones that Rockcraft currently does.

Contrast this with [Snapcraft](https://canonical-snapcraft.readthedocs-hosted.com/en/latest/explanation/remote-build/#open-vs-closed-source): there users can do a bit more work to do remote builds for private projects.

How about something longer like "...because the remote builds triggered by Rockcraft are publicly available"?

Author: ^did this

Reviewer: :+1:

> - All architectures defined on the project's ``platforms`` entry - either
>   explicitly via the ``build-on`` key or implicitly through the platform
>   shorthand - will be built. There is currently no way to restrict the set of

Reviewer: Sphinx parses double-hyphens as en dashes (–).

```suggestion
- All architectures defined in the project's ``platforms`` entry -- either
  explicitly through the ``build-on`` key or implicitly through the platform
  shorthand -- are built. There's currently no way to restrict the set of
```

>   shorthand - will be built. There is currently no way to restrict the set of

Reviewer: "platform shorthand" might be difficult for newer users to intuit.

Author: You're right, and it's a bit too much detail for this point anyway. Reworded this entry.
